federatedscope/contrib/trainer/torch_example.py

```diff
@@ -0,0 +1,101 @@
+import inspect
+from federatedscope.register import register_trainer
+from federatedscope.core.trainers import BaseTrainer
+
+# An example for converting torch training process to FS training process
+
+# Refer to `federatedscope.core.trainers.BaseTrainer` for interface.
+
+# Try with FEMNIST:
+#  python federatedscope/main.py --cfg scripts/example_configs/femnist.yaml \
+#  trainer.type mytorchtrainer federate.sample_client_rate 0.01 \
+#  federate.total_round_num 5 eval.best_res_update_round_wise_key test_loss
+
+
+class MyTorchTrainer(BaseTrainer):
+    def __init__(self, model, data, device, **kwargs):
+        import torch
+        # NN modules
+        self.model = model
+        # FS `ClientData` or your own data
+        self.data = data
+        # Device name
+        self.device = device
+        # kwargs
+        self.kwargs = kwargs
+        # Criterion & Optimizer
+        self.criterion = torch.nn.CrossEntropyLoss()
+        self.optimizer = torch.optim.SGD(self.model.parameters(),
+                                         lr=0.001,
+                                         momentum=0.9,
+                                         weight_decay=1e-4)
+
+    def train(self):
+        # _hook_on_fit_start_init
+        self.model.to(self.device)
+        self.model.train()
+
+        total_loss = num_samples = 0
+        # _hook_on_batch_start_init
+        for x, y in self.data['train']:
+            # _hook_on_batch_forward
+            x, y = x.to(self.device), y.to(self.device)
+            outputs = self.model(x)
+            loss = self.criterion(outputs, y)
+
+            # _hook_on_batch_backward
+            self.optimizer.zero_grad()
```

Comment:
I suggest putting this line before the forward process (i.e., `outputs = self.model(x)`).

Reply:
If so, should we modify [GeneralTorchTrainer](https://github.com/alibaba/FederatedScope/blob/45f2da666d5c3f5c47d485124c19383e49acdd7a/federatedscope/core/trainers/torch_trainer.py#L263), which I follow the principle of.